src/jit/morph.cpp

```diff
@@ -13983,39 +13983,62 @@ GenTree* Compiler::fgMorphSmpOpOptional(GenTreeOp* tree)
     if (fgGlobalMorph && (oper == GT_ADD) && !tree->gtOverflow() && (op1->gtOper == GT_ADD) && !op1->gtOverflow() &&
         varTypeIsIntegralOrI(typ))
     {
+        GenTree* ad1 = op1->gtOp.gtOp1;
         GenTree* ad2 = op1->gtOp.gtOp2;
 
         if (op2->OperIsConst() == 0 && ad2->OperIsConst() != 0)
         {
-            // This takes
-            //       + (tree)
-            //      / \.
-            //     /   \.
-            //    /     \.
-            //   + (op1) op2
-            //  / \.
-            //     \.
-            //     ad2
+            //  This takes
+            //        + (tree)
+            //       / \.
+            //      /   \.
+            //     /     \.
+            //    + (op1) op2
+            //   / \.
+            //  /   \.
+            // ad1  ad2
             //
-            // And it swaps ad2 and op2.  If (op2) is varTypeIsGC, then this implies that (tree) is
-            // varTypeIsGC.  If (op1) is not, then when we swap (ad2) and (op2), then we have a TYP_INT node
-            // (op1) with a child that is varTypeIsGC.  If we encounter that situation, make (op1) the same
-            // type as (tree).
-            //
-            // Also, if (ad2) is varTypeIsGC then (tree) must also be (since op1 is), so no fixing is
-            // necessary
+            // and it swaps ad2 and op2.
+
+            var_types typ2   = op2->TypeGet();
+            var_types ad1typ = ad1->TypeGet();
+
+            bool doTransformation = true;
 
-            if (varTypeIsGC(op2->TypeGet()))
+            // Don't create a byref pointer that may point outside of the ref object. If a GC happens, the byref won't
+            // get updated. This can happen, if one of the int components is negative.
+            // It also requires the address generation be in a fully-interruptible code region.
+            // We allow the transformation only if the constant is non-negative and the other int operand is unsigned.
```

Comment:
The unsigned case is probably not going to happen. If one operand is GC then the others are supposed to be pointer sized integers and these are not normally unsigned in JIT's IR (e.g. you can have a `TYP_USHORT` indir but not a `TYP_UINT`/`TYP_ULONG` one).

Reply:
I think you are right, I wasn't able to construct an example with an unsigned case so I simplified the fix.